dipy/nn/torch/evac.py

```diff
@@ -0,0 +1,551 @@
+#!/usr/bin/python
+"""
+Class and helper functions for fitting the EVAC+ model.
+"""
+
+import logging
+
+import numpy as np
+
+from dipy.align.reslice import reslice
+from dipy.data import get_fnames
+from dipy.nn.utils import (
+    normalize,
+    recover_img,
+    set_logger_level,
+    transform_img,
+)
+from dipy.segment.utils import remove_holes_and_islands
+from dipy.testing.decorators import doctest_skip_parser, warning_for_keywords
+from dipy.utils.deprecator import deprecated_params
+from dipy.utils.optpkg import optional_package
+
+torch, have_torch, _ = optional_package("torch", min_version="2.2.0")
+if have_torch:
+    from torch.nn import (
+        Conv3d,
+        ConvTranspose3d,
+        Dropout3d,
+        LayerNorm,
+        Module,
+        ModuleList,
+        ReLU,
+        Softmax,
+    )
+else:
+
+    class Module:
+        pass
+
+    logging.warning(
+        "This model requires Pytorch.\
+                    Please install these packages using \
+                    pip."
+    )
+
+logging.basicConfig()
+logger = logging.getLogger("EVAC+")
+
+
+def prepare_img(image):
+    """
+    Function to prepare image for model input
+    Specific to EVAC+
+
+    Parameters
+    ----------
+    image : np.ndarray
+        Input image
+
+    Returns
+    -------
+    input_data : dict
+    """
+    input1 = np.moveaxis(image, -1, 0)
+    input1 = np.expand_dims(input1, 1)
+
+    input2, _ = reslice(image, np.eye(4), (1, 1, 1), (2, 2, 2))
+    input2 = np.moveaxis(input2, -1, 0)
+    input2 = np.expand_dims(input2, 1)
+
+    input3, _ = reslice(image, np.eye(4), (1, 1, 1), (4, 4, 4))
+    input3 = np.moveaxis(input3, -1, 0)
+    input3 = np.expand_dims(input3, 1)
+
+    input4, _ = reslice(image, np.eye(4), (1, 1, 1), (8, 8, 8))
+    input4 = np.moveaxis(input4, -1, 0)
+    input4 = np.expand_dims(input4, 1)
+
+    input5, _ = reslice(image, np.eye(4), (1, 1, 1), (16, 16, 16))
+    input5 = np.moveaxis(input5, -1, 0)
+    input5 = np.expand_dims(input5, 1)
+
+    input_data = [
+        torch.from_numpy(input1).float(),
+        torch.from_numpy(input2).float(),
+        torch.from_numpy(input3).float(),
+        torch.from_numpy(input4).float(),
+        torch.from_numpy(input5).float(),
+    ]
+
+    return input_data
+
+
+class MoveDimLayer(Module):
+    def __init__(self, source_dim, dest_dim):
+        super(MoveDimLayer, self).__init__()
+        self.source_dim = source_dim
+        self.dest_dim = dest_dim
+
+    def forward(self, x):
+        return torch.movedim(x, self.source_dim, self.dest_dim)
+
+
+class ChannelSum(Module):
+    def __init__(self):
+        super(ChannelSum, self).__init__()
+
+    def forward(self, inputs):
+        return torch.sum(
+            inputs, dim=1, keepdim=True
+        )  # Summing across channel dimension
```

Comment:
why not function documentation instead comment ?

Reply:
It's from using llms to transfer model architectures. Unnecessary so removed.